### PR TITLE
Issue 2 mono api

### LINKED ISSUE #2: Implement MonoAPI
- [x] Implement Monobank API that fulfills our needs, in particular:
  - -[x] Receiving information about transactions (happened after some date);
  - -[x] Receiving information about user cards/accounts;
  - -[x] Receiving information about user balance.
  - -[x] Implement [Bank DataSource](https://github.com/AlmostGreatBand/lemon-server/blob/main/src/datasources/BankDataSource.js)

If Monobank API doesn't support some of this requests, write it here or in Slack.

### PR DESCRIPTION
**Pull request by NeProgramist**
close #2

### CODE REVIEW

File: mono-api/mono-api.js [hunk 1]
```diff
@@ -0,0 +1,74 @@
+'use strict'; 
+
+const http = require('https');
+const BASE_URL = 'api.monobank.ua';
+/*
+    this token should be passed from repository layer, it's here just for tests
+*/ 
+const TOKEN = 'uMLmvCFm6kA4N8mrflxayXMcZPqFUrxpm9q_CxBsMkaY';
+
+function createRequestBody(path, token) {
```
Comment (ALEGATOR1209) [inline mono-api/mono-api.js]: But this function creates not only body, but headers also, doesn't it?

File: mono-api/mono-api.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        headers: {
+            'X-Token': token,
+            'Content-Type': 'application/json; charset=UTF-8',
+        }
+    };
+}
+
+function getData(path, token) {
+    const options = createRequestBody(path, token);
+    const req = http.request(options);
```
Comment (ALEGATOR1209) [inline mono-api/mono-api.js]: Maybe, it'd be better something like this:

```
const req = createRequest(path, token);
```

File: mono-api/mono-api.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    req.on('error', err => {
+        try {
+            console.log(err.toString());
+            const e = JSON.parse(err);
+            console.log(`Error: ${e.errorDescription}`);
+            return e;
+        } catch(e) {
+            console.log(`Error: something went wront while trying to parse err: ${e}`);
+        }
+    });
```
Comment (ALEGATOR1209) [inline mono-api/mono-api.js]: Maybe, move it to function like `networkErrorHandler`?

File: mono-api/mono-api.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            console.log(`Error: something went wront while trying to parse err: ${e}`);
+        }
+    }); 
+    req.end();
+}
+
+function getTransactions(token, account, from, to) {
+    let queryString = `/personal/statement/${account}/${from}`;
+    if (to) queryString += `/${to}`;
+    if (account !== undefined && from !== undefined) return getData(queryString, token);
```
Comment (ALEGATOR1209) [inline mono-api/mono-api.js]: Is parameter `from` required by Monobank API? If not, we could make it optional like `to`.

**Review by ALEGATOR1209 — CHANGES_REQUESTED** [review]
Resolve mentioned issues + implement our data source interface

File: mono-api/mono-api.js (hunk 4 above)
Comment (NeProgramist) [reply mono-api/mono-api.js]: `account` and `to` are optional, `from` isn't
but there is no situation when we don't pass the `account` into this function, so in our case `from` and `account` are mandatory

File: mono-api/mono-api.js (hunk 3 above)
Comment (NeProgramist) [reply mono-api/mono-api.js]: I thought that `networkErrorHandler` would work just with network error, not with mono-API errors
I don't think that `networkErrorHandler` should know about API errors

Comment (ALEGATOR1209) [reply mono-api/mono-api.js]: Create another function, like `apiErrorHandler`?

File: mono-api/mono-api.js (hunk 1 above)
Comment (NeProgramist) [reply mono-api/mono-api.js]: ok, I'll change it to `createRequest`

File: mono-api/mono-api.js (hunk 2 above)
Comment (NeProgramist) [reply mono-api/mono-api.js]: Should I move `http.request` to `createRequest` funciton?

File: mono-api/mono-api.js (hunk 3 above)
Comment (NeProgramist) [reply mono-api/mono-api.js]: we talked about it with @Hud-Volodymyr, he said that's repository module responsibility to handle API errors
and mono API has just one field in error-model, that's `errorDescription`, so I don't know is it worth it or not

**Review by ALEGATOR1209 — CHANGES_REQUESTED** [review]
Why delete hooks and CI?

**Comment by ALEGATOR1209** [comment]
Also remove usage example to separate file and don't publish your own token to GitHub.
@NeProgramist

File: src/datasources/mono/mono.js [hunk 5]
```diff
@@ -1,3 +1,101 @@
...
+            headers: {
+                'X-Token': this.token,
+                'Content-Type': 'application/json; charset=UTF-8',
+            }
+        };
+        return http.request(options);
+    }
+
+    #getData = async (path) => {
+        console.log(this.token);
```
Comment (Hud-Volodymyr) [inline src/datasources/mono/mono.js]: Remove console.log

File: src/datasources/mono/mono.js [hunk 6]
```diff
@@ -1,3 +1,101 @@
...
+        const req = this.#makeRequest(path);
+        console.log(req);
```
Comment (Hud-Volodymyr) [inline src/datasources/mono/mono.js]: No console logs in master code

File: src/datasources/mono/mono.js [hunk 7]
```diff
@@ -1,3 +1,101 @@
...
+            } else {
+                let stream = '';
+                res.on('data', chunk => {
+                    stream += chunk;
+                });
+
+                res.on('end', () => {
+                    try {
+                        const data = JSON.parse(stream);
+                        console.log(data);
```
Comment (Hud-Volodymyr) [inline src/datasources/mono/mono.js]: Unnecessary console.log

File: src/datasources/mono/mono.js [hunk 8]
```diff
@@ -1,3 +1,101 @@
...
+                        console.log(errMessage);
+                        throw new Error(errMessage);
+                    }
+                })
+            }
+        });
+    
+        req.on('error', err => {
+            try {
+                console.log(err.toString());
```
Comment (Hud-Volodymyr) [inline src/datasources/mono/mono.js]: Unnecessary console.log

**Review by Hud-Volodymyr — CHANGES_REQUESTED** [review]

**Comment by NeProgramist** [comment]
> Also remove usage example to separate file and don't publish your own token to GitHub.
> @NeProgramist

I publish my token in order to make it possible for you to test it on your own, but ok, I'll do it

**Comment by NeProgramist** [comment]
> Why delete hooks and CI?

because our hooks only for Unix-based OS, unfortunately, I have Windows so I cannot push any changes to git

File: src/datasources/mono/mono.js [hunk 9]
```diff
@@ -1,3 +1,90 @@
...
+
+    async getTransactions(account, from, to) {
+        let queryString = `${MonoDataSource.#ROUTES.transactions}/${account}/${from.getTime()}`;
+        if (to) queryString += `/${to.getTime()}`;
+
+        if (account !== undefined && from !== undefined) {
+            return this.#getData(queryString);
+        } else {
+            const errMessage = 'Error: account or from-date is not specified';
+            console.log(errMessage);
```
Comment (ALEGATOR1209) [inline src/datasources/mono/mono.js]: Maybe, it's better to use console.error?

Comment (ALEGATOR1209) [reply src/datasources/mono/mono.js]: In this and other places below

File: src/datasources/mono/mono_usage.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+const MonoDataSource = require('./mono.js')
+const monoDS = new MonoDataSource(process.env.MONO_TOKEN);
+(async () => {
+    try {
+        const transactions = await monoDS.getTransactions(undefined, new Date(2020, 10, 15), new Date(2020, 10, 26));
```
Comment (ALEGATOR1209) [inline src/datasources/mono/mono_usage.js]: Leave some comment that it's desirable behavior or something, it's a bit confusing that you pass undefined here. Maybe, add some successfull example too.

**Review by ALEGATOR1209 — CHANGES_REQUESTED** [review]
Still dislike that you deleted webhooks and actions config file.

If you have troubles with using hooks, you can change or delete them locally without pushing it to GitHub.

File: src/datasources/mono/mono_usage.js (hunk 10 above)
Comment (NeProgramist) [reply src/datasources/mono/mono_usage.js]: ok

**Review by ALEGATOR1209 — DISMISSED** [review]

**Review by ALEGATOR1209 — APPROVED** [review]
:+1:

**Review by Hud-Volodymyr — APPROVED** [review]
Looks good

**Pull request merged**